src/metabase/models/params/custom_values.clj

```diff
@@ -0,0 +1,101 @@
+(ns metabase.models.params.custom-values
+  (:require
+    [clojure.string :as str]
+    [metabase.mbql.normalize :as mbql.normalize]
+    [metabase.models :refer [Card]]
+    [metabase.query-processor :as qp]
+    [metabase.search.util :as search]
+    [metabase.util.i18n :refer [tru]]
+    [metabase.util.schema :as su]
+    [schema.core :as s]
+    [toucan.db :as db]))
+
+;; -------------------------------------- source=card --------------------------------------
+
+(def ^{:dynamic true
+       :doc     "Maximum number of rows returned when running a card."}
+  *max-rows* 2000)
+
+(defn- field-ref->mbql-field
+  "Find field by field_ref.
+
+  If field_ref is a:
+  - field: returns itself
+  - aggregation: returns a field using name [:field aggregation-name nil]"
+  [field-ref result-metadata card-id]
+  (let [field (first (filter #(= (:field_ref %) field-ref) result-metadata))]
+    (when-not field
+      (throw (ex-info (tru "No matching field found")
+                      {:status-code 400
+                       :card-id     card-id
+                       :field-ref   field-ref})))
+    (case (first field-ref)
+      :aggregation
+      [:field (:name field) {:base-type ((some-fn :effective_type :base_type) field)}]
+
+      :field
+      field-ref
```

Comment:
Unless this is done on the frontend, I think you'll have to strip out :binning and :temporal-unit from the field here. Ideally we'd have a universal way to key columns that didn't include this information, and it should probably live in a mbql shared namespace so it can be called outside the query-processor namespaces. 

See this for the existing logic that's used in the query processor
https://github.com/metabase/metabase/blob/464ec9e5bede2aec50391d50f8f7b5b600e61f96/src/metabase/query_processor/middleware/add_implicit_clauses.clj#L58

Reply:
I found this discussion between Cam and Dan that might give more context.
https://github.com/metabase/metabase/pull/19158#r770039976

Note this is an existing issue with `field-ref->key`, so don't look to copy that function.

Reply:
that's a good point.

Given the inherent issue with identifying a key, I wonder if it's better to let UI construct a `field` so that we could directly use it in the mbql. I believe UI already did this because we could select any fields to be returned from a card/table in the Question editor.


This way we don't have to deal with field finding and the `binning, temporal-unit` options

Reply:
good news, we have finally got around and let UI construct the field 🎉